### PR TITLE
Parallel UDF optimizations

### PR DESCRIPTION
**Pull request by dreadatour**
"Parallel" mode is not optimal now. It took more time and much more memory (script can even be killed by OOMKiller). In this PR we are going to profile and optimize parallel mode.

---

### Before optimization

Test dataset with `109662` rows:
```
$ ls -l dvcx-datacomp-small/metadata-small
total 1315952
-rw-r--r--@ 1 vlad  staff  673764696 Jul 24 17:15 036d6b9ae87a00e738f8fc554130b65b.npz
$
```

#### Single process mode

Test query `laion_emb.py`:
```python
from datachain.lib.dc import DataChain
from datachain.lib.webdataset_laion import process_laion_meta

(
    DataChain.from_storage("dvcx-datacomp-small/metadata-small/")
    .gen(emd=process_laion_meta)
    .map(stem=lambda file: file.get_file_stem(), params=["emd.file"], output=str)
    .save("laion_emb")
)
```

##### Run time

```
$ time python laion_emb.py
Processed: 1 rows [00:00, 2532.79 rows/s]
Download: 643MB [00:53, 12.7MB/s]s]
Processed: 1 rows [00:53, 53.10s/ rows]
Generated: 109662 rows [00:52, 2071.59 rows/s]
Processed: 109662 rows [00:18, 6039.80 rows/s]
python laion_emb.py  58.26s user 37.74s system 62% cpu 2:32.64 total
$
```

##### Memory consumption

Pasted image 20240806224139

#### Parallel mode

Test query `laion_emb_parallel.py`:

```python
from datachain.lib.dc import DataChain
from datachain.lib.webdataset_laion import process_laion_meta

(
    DataChain.from_storage("dvcx-datacomp-small/metadata-small/")
    .settings(cache=True, parallel=4)
    .gen(emd=process_laion_meta)
    .map(stem=lambda file: file.get_file_stem(), params=["emd.file"], output=str)
    .save("laion_emb")
)
```

##### Run time

```
$ time python laion_emb_parallel.py
Processed: 1 rows [00:00, 2236.96 rows/s]
Processed: 1 rows [00:34, 34.10s/ rows]
Generated: 109662 rows [00:34, 3216.25 rows/s]
Processed: 109662 rows [03:54, 467.85 rows/s]
python laion_emb_parallel.py  372.29s user 61.47s system 84% cpu 8:30.86 total
$
```

##### Memory consumption

Pasted image 20240806225238

#### Comparing

```
 Action     | Single process        | Parallel              | Difference
------------+-----------------------+-----------------------+------------
 Generated  | 00:52, 2071.59 rows/s | 00:34, 3216.25 rows/s | 65%
 Processed  | 00:18, 6039.80 rows/s | 03:54,  467.85 rows/s | 1300%
 Total time | 02:32                 | 08:30                 | 336%
 Max memory | 3,2 Gb                | 15 Gb                 | 469%
```
Parallel mode is three times slower and take 4-5 times more memory.

#### Profiling

Let's profile "processing" step:

```
.map(stem=lambda file: file.get_file_stem(), params=["emd.file"], output=str)
```

#### Main process

```
         4409226518 function calls (3759456933 primitive calls) in 720.459 seconds

   Ordered by: cumulative time

   ncalls         tottime  percall cumtime  percall filename:lineno(function)
109665/109663      1.638    0.000 1643.795    0.015 /Users/vlad/work/iterative/datachain/src/datachain/query/dispatch.py:262(run_udf_parallel)
109667/109666      0.423    0.000  760.557    0.007 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/reduction.py:51(dumps)
109675/109674      0.081    0.000  727.287    0.007 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/dill/_dill.py:418(dump)
323616022/109674  68.524    0.000  708.879    0.006 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/dill/_dill.py:367(save)
109675/109674      0.179    0.000  708.331    0.006 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:470(dump)
323616022/109674 247.330    0.000  705.380    0.006 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:529(save)
2412816/109670     2.871    0.000  689.873    0.006 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:615(save_reduce)
109830/109670      0.646    0.000  664.102    0.006 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:970(_batch_setitems)
   548358          0.586    0.000  661.603    0.001 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:919(save_list)
   548358         29.646    0.000  630.127    0.001 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:932(_batch_appends)
109699/109698      0.290    0.000  412.868    0.004 {built-in method builtins.next}
308808192         63.378    0.000  146.392    0.000 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:776(save_float)
109664/109663      2.066    0.000   90.228    0.001 /Users/vlad/work/iterative/datachain/src/datachain/data_storage/warehouse.py:187(dataset_select_paginated)
323725696         46.299    0.000   68.786    0.000 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:211(commit_frame)
334694224         40.225    0.000   66.634    0.000 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:235(write)
   109666         0.056     0.000   63.358    0.001 /Users/vlad/work/iterative/datachain/src/datachain/query/dispatch.py:58(get_from_queue)
   219324         0.119     0.000   48.065    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/fsspec/callbacks.py:112(relative_update)
   109666         0.057     0.000   43.746    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/queues.py:137(get_nowait)
   109666         0.198     0.000   33.116    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/queues.py:101(get)
   109662         0.069     0.000   27.632    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/fsspec/callbacks.py:309(call)
326796653        27.490     0.000   27.490    0.000 {method 'get' of 'dict' objects}
335132904        25.960     0.000   25.960    0.000 {method 'write' of '_io.BytesIO' objects}
2413072/2193268   2.039     0.000   24.792    0.000 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:868(save_tuple)
315828804        21.915     0.000   21.915    0.000 {built-in method _struct.pack}
323835370        21.769     0.000   21.769    0.000 {method 'tell' of '_io.BytesIO' objects}
334474033        20.220     0.000   20.220    0.000 {built-in method builtins.getattr}
321421742        19.009     0.000   20.129    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/dill/_dill.py:316(get)
348075682        18.344     0.000   18.344    0.000 {built-in method builtins.id}
   109662         0.063     0.000   18.231    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/tqdm/std.py:1198(update)
       11         0.007     0.001   16.315    1.483 /Users/vlad/work/iterative/datachain/src/datachain/data_storage/sqlite.py:666(insert_rows)
   109666         0.103     0.000   15.016    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/connection.py:256(poll)
   109666         0.131     0.000   13.051    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/connection.py:211(recv_bytes)
323616022        12.606     0.000   12.982    0.000 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:599(persistent_id)
       14         0.022     0.002   12.723    0.909 /Users/vlad/work/iterative/datachain/src/datachain/data_storage/sqlite.py:69(wrapper)
   658072         1.167     0.000   12.602    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/dill/_dill.py:1750(save_type)
     6244         0.012     0.000   12.444    0.002 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/tqdm/std.py:1325(refresh)
   109666         0.150     0.000   12.357    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/connection.py:432(_recv_bytes)
     6245         0.027     0.000   11.388    0.002 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/tqdm/std.py:1464(display)
   109666         0.150     0.000    9.715    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/connection.py:185(send_bytes)
  9542462         5.320     0.000    9.456    0.000 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:485(memoize)
  3510012         2.393     0.000    7.910    0.000 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:845(save_str)
   109666         0.138     0.000    7.589    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/connection.py:442(_poll)
   658092         1.060     0.000    7.583    0.000 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:1048(save_global)
     6245         0.043     0.000    6.994    0.001 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/tqdm/std.py:457(print_status)
   548310         0.257     0.000    6.839    0.000 /Users/vlad/work/iterative/datachain/src/datachain/sql/sqlite/types.py:34(convert_array)
   548310         6.582     0.000    6.582    0.000 {orjson.loads}
   109666         0.304     0.000    6.496    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/connection.py:1125(wait)
     6245         0.041     0.000    5.556    0.001 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/tqdm/std.py:451(fp_write)
```

From `720` total seconds:

- `69 sec` is `dill.save`
- `247 sec` is `pickle.save`
- `30 sec` is `pickle._batch_appends`
- `63 sec` is `pickle.save_float`
- `46 sec` is `pickle.commit_frame`
- `40 sec` is `pickle.write`
- `168` more seconds on other conversion calls

It is ~`660` seconds summarized or ~`92%` of time for types conversion and pickling.

##### Subprocess (parallel worker)

```
         20331539 function calls (19619464 primitive calls) in 719.039 seconds

   Ordered by: cumulative time

   ncalls    tottime  percall  cumtime  percall filename:lineno(function)
27536/27535    0.397    0.000  756.787    0.027 /Users/vlad/work/iterative/datachain/src/datachain/lib/udf.py:43(run)
27536/27535    0.021    0.000  750.928    0.027 /Users/vlad/work/iterative/datachain/src/datachain/query/dispatch.py:419(get_inputs)
    27535      0.151    0.000  750.903    0.027 /Users/vlad/work/iterative/datachain/src/datachain/query/dispatch.py:58(get_from_queue)
    57478    462.631    0.008  462.631    0.008 {built-in method time.sleep}
    25085      0.043    0.000  143.502    0.006 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/connection.py:185(send_bytes)
    25085      0.031    0.000  143.454    0.006 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/connection.py:409(_send_bytes)
    25085      0.028    0.000  143.408    0.006 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/connection.py:384(_send)
137634/53453 140.569    0.001  140.512    0.003 {method 'acquire' of '_multiprocessing.SemLock' objects}
    25085    107.034    0.004  107.034    0.004 {built-in method posix.write}
    85013      0.027    0.000    4.398    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/queues.py:137(get_nowait)
    85013      0.106    0.000    4.371    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/queues.py:101(get)
    25086      0.071    0.000    2.290    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/reduction.py:51(dumps)
    25086      0.017    0.000    2.020    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/dill/_dill.py:418(dump)
    70003      0.044    0.000    1.987    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/multiprocess/connection.py:256(poll)
    25086      0.047    0.000    1.983    0.000 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:470(dump)
    27534      0.068    0.000    1.949    0.000 /Users/vlad/work/iterative/datachain/src/datachain/lib/udf.py:65(run_once)
    27535      0.014    0.000    1.872    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/dill/_dill.py:291(loads)
    27535      0.061    0.000    1.858    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/dill/_dill.py:283(load)
301032/25086   0.110    0.000    1.816    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/dill/_dill.py:367(save)
301032/25086   0.359    0.000    1.793    0.000 /opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pickle.py:529(save)
    27535      0.019    0.000    1.738    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/dill/_dill.py:443(load)
    27535      1.115    0.000    1.716    0.000 {function Unpickler.load at 0x103436ca0}
50172/25086    0.086    0.000    1.710    0.000 /Users/vlad/.virtualenvs/datachain/lib/python3.12/site-packages/dill/_dill.py:1195(save_module_dict)
    27534      0.065    0.000    1.704    0.000 /Users/vlad/work/iterative/datachain/src/datachain/lib/udf.py:225(__call__)
```

From `719` total seconds:

- `463 sec` is `time.sleep` (waiting for a data in input queue using `queue.get_nowait`)
- `141 sec` is acquiring `_multiprocessing.SemLock`
- `107 sec` is writing to output queue

### Optimization

From these results it is clear the problem is in pickling/types conversion and writing/reading to/from multiprocessing queues.

Taking this steps to optimize this:
- return rows instead of dicts from SQL query (reduce memory consumption)
- marshal UDF rows with msgpack instead of dill (speedup pickling/types conversion)
- add batching to UDF results (speedup writing/reading to/from multiprocessing queues)

##### Run time

```
$ time python laion_emb.py
Listing file:///: 1 objects [00:00, 378.34 objects/s]
Processed: 1 rows [00:00, 2864.96 rows/s]
Download: 643MB [00:42, 15.9MB/s]s]
Processed: 1 rows [00:42, 42.45s/ rows]
Generated: 109662 rows [00:42, 2594.06 rows/s]
Processed: 109662 rows [00:14, 7325.12 rows/s]
python laion_emb.py  55.90s user 36.86s system 73% cpu 2:05.88 total
$
```

##### Memory consumption

Pasted image 20240806233046

##### Comparison

Difference with "before optimization" (just to be sure we don't break anything):

```
 Action     | Before                | After                 | Difference
------------+-----------------------+-----------------------+------------
 Generated  | 00:52, 2071.59 rows/s | 00:42, 2594.06 rows/s | 81%
 Processed  | 00:18, 6039.80 rows/s | 00:14, 7325.12 rows/s | 78%
 Total time | 02:32                 | 02:06                 | 83%
 Max memory | 3,2 Gb                | 3,2 Gb                | same
```

~`80%` faster after optimization with same amount of memory.

##### Run time

```
$ time python laion_emb_parallel.py
Listing file:///: 1 objects [00:00, 343.18 objects/s]
Processed: 1 rows [00:00, 2866.92 rows/s]
Processed: 1 rows [00:02,  2.30s/ rows]rows/s]
Generated: 109662 rows [00:45, 2434.31 rows/s]
Processed: 109662 rows [00:20, 5329.11 rows/s]
python laion_emb_parallel.py  96.14s user 63.98s system 126% cpu 2:06.77 total
(datachain) ~/work/iterative/playground/laion
```

##### Memory consumption

Pasted image 20240806234514

##### Comparison

Difference with "before optimization" (just to be sure we don't break anything):

```
 Action     | Before                | After                 | Difference
------------+-----------------------+-----------------------+------------
 Generated  | 00:34, 3216.25 rows/s | 00:45, 2434.31 rows/s | 132%
 Processed  | 03:54,  467.85 rows/s | 00:20, 5329.11 rows/s | 9%
 Total time | 08:30                 | 02:07                 | 25%
 Max memory | 15 Gb                 | 3,6 Gb                | 23%
```

4 times faster and 4 times less memory.
"Processing" step was 13 times slower than in "single process" mode, now it is 10 times faster than before.

### Result

Overall now total time and max memory consumption is about the same as in single process mode.

For CPU-intensive jobs difference should be more visible.

### CODE REVIEW

File: src/datachain/query/dispatch.py [hunk 1]
```diff
@@ -44,44 +53,6 @@ def get_n_workers_from_arg(n_workers: Optional[int] = None) -> int:
...
-# For more context on the get_from_queue and put_into_queue functions, see the
-# discussion here:
-# https://github.com/iterative/dvcx/pull/1297#issuecomment-2026308773
-# This problem is not exactly described by, but is also related to these Python issues:
-# https://github.com/python/cpython/issues/66587
-# https://github.com/python/cpython/issues/88628
-# https://github.com/python/cpython/issues/108645
-
-
-def get_from_queue(queue: Queue) -> Any:
```
Comment (shcheklein) [inline src/datachain/query/dispatch.py:56]: are we moving things around in this case? can it be done please as a separate PR if that's the case? (PR is large, quite nuanced any reduction in changes can help to review it).

File: src/datachain/data_storage/warehouse.py [hunk 2]
```diff
@@ -200,23 +200,17 @@ def dataset_row_cls(self) -> type["DataTable"]:
...
+        paginated_query = query.limit(page_size)
 
-        if not order_by:
-            ordering = [cols.sys__id]
-        else:
-            ordering = order_by  # type: ignore[assignment]
-
-        # reset query order by and apply new order by id
-        paginated_query = query.order_by(None).order_by(*ordering).limit(page_size)
+        if not paginated_query._order_by_clauses:
```
Comment (shcheklein) [inline src/datachain/data_storage/warehouse.py:211]: is there any improvement in this or it's a more or less cosmetic?

File: src/datachain/data_storage/warehouse.py [hunk 3]
```diff
@@ -235,7 +229,7 @@ def dataset_select_paginated(
                 processed = False
                 for row in results:
                     processed = True
-                    yield RowDict(zip(cols_names, row))
```
Comment (shcheklein) [inline src/datachain/data_storage/warehouse.py:238]: is it the main change here? (avoiding creating dicts?)

File: src/datachain/utils.py [hunk 4]
```diff
@@ -251,6 +251,21 @@ def batched(iterable: Iterable[_T_co], n: int) -> Iterator[tuple[_T_co, ...]]:
...
+    # batched('ABCDEFG', 3) --> ABC DEF G
+    if n < 1:
+        raise ValueError("Batch size must be at least one")
+    it = iter(iterable)
+    while True:
+        chunk_it = islice(it, n)
+        try:
+            first_el = next(chunk_it)
+        except StopIteration:
+            return
```
Comment (shcheklein) [inline src/datachain/utils.py:265]: could you remind me - does it mean it can return None? should it be also throwing StopIteration (to behave like an iterator I guess)

File: src/datachain/lib/udf.py [hunk 5]
```diff
@@ -35,42 +37,47 @@ class UDFAdapter(_UDFBase):
     def __init__(
         self,
         inner: "UDFBase",
-        properties: UDFProperties,
+        properties: "UDFProperties",
```
Comment (shcheklein) [inline src/datachain/lib/udf.py:40]: minor: seems this change is not needed?

Comment (shcheklein) [reply src/datachain/lib/udf.py:40]: or you should move UDFProperties into the `TYPE_CHECKING` section? (overall, same ask - can we do less minor updates or split them if they are not essential to the PR, thanks!)

File: src/datachain/lib/udf.py [hunk 6]
```diff
@@ -35,42 +37,47 @@ class UDFAdapter(_UDFBase):
...
-        udf_inputs: "Iterable[BatchingResult]",
+        udf_fields: "Sequence[str]",
+        udf_inputs: "Iterable[RowsInput]",
         catalog: "Catalog",
         is_generator: bool,
         cache: bool,
         download_cb: Callback = DEFAULT_CALLBACK,
         processed_cb: Callback = DEFAULT_CALLBACK,
-    ) -> Iterator[Iterable["UDFResult"]]:
+    ) -> "Iterator[Iterable[UDFResult]]":
```
Comment (shcheklein) [inline src/datachain/lib/udf.py:54]: is it an essential change or cosmetics?

File: src/datachain/query/dispatch.py [hunk 7]
```diff
@@ -348,7 +330,7 @@ def run_udf_parallel(  # noqa: C901, PLR0912
 
 
 class WorkerCallback(Callback):
-    def __init__(self, queue: multiprocess.Queue):
+    def __init__(self, queue: "multiprocess.queues.Queue"):
```
Comment (rlamy) [inline src/datachain/query/dispatch.py:333]: `Queue` is documented as being part of `multiprocess`. AFAICT, `multiprocess.queues` is considered an implementation detail.

**Review by dmpetrov — APPROVED** [review]
Thank you, Vlad! Great improvements.

Some high-level comments:

>> return rows instead of dicts from SQL query (reduce memory consumption)

Good idea!

>> marshal UDF rows with msgpack instead of dill

1. **Is there a way to avoid the conversion?** We use a very limited subset of types. Why not just pass unflattened objects as tuple?
2. **Is there a way to avoid marshaling whatsoever?** We can write directly to DB.

Think of these as ideas for the next steps.



>>> add batching to UDF results

Great!

**Review by dtulga — APPROVED** [review]
LGTM, thanks for the improvements! I don't have any other new comments (other than duplicates of existing comments).

File: src/datachain/query/dataset.py [hunk 8]
```diff
@@ -519,6 +519,7 @@ def populate_udf_table(self, udf_table: "Table", query: Select) -> None:
                     udf = self.udf
 
                 warehouse = self.catalog.warehouse
+                udf_fields = [str(c.name) for c in query.selected_columns]
```
Comment (shcheklein) [inline src/datachain/query/dataset.py:522]: Q: can you please explain / copy paste from the description the relevant part - what is the deal with these `udf_fields`. I'm a bit concerned with code like:

`def _create_worker(self, udf_fields: Sequence[str])`

seems weird that worker needs this (and only this)

File: src/datachain/query/batch.py [hunk 9]
```diff
@@ -28,9 +36,9 @@ class BatchingStrategy(ABC):
     @abstractmethod
     def __call__(
         self,
-        execute: Callable,
-        query: sa.sql.selectable.Select,
-    ) -> Generator[BatchingResult, None, None]:
+        execute: Callable[..., Generator[Sequence, None, None]],
+        query: "Select",
+    ) -> Generator[RowsInput, None, None]:
```
Comment (shcheklein) [inline src/datachain/query/batch.py:41]: Q: should the name be RowsOutput or RowsResult?

File: src/datachain/query/batch.py [hunk 10]
```diff
@@ -59,31 +67,24 @@ def __init__(self, count: int):
...
         page_size = math.ceil(SELECT_BATCH_SIZE / self.count) * self.count
 
         # select rows in batches
-        results: list[RowDict] = []
-
-        with contextlib.closing(
-            execute(
-                query,
-                page_size=page_size,
-                limit=query._limit,
```
Comment (shcheklein) [inline src/datachain/query/batch.py:75]: why is it safe / where do we now do limit / order by?

File: src/datachain/query/queue.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        )
+
+    raise TypeError(f"Unknown type: {obj}")
+
+
+def msgpack_pack(obj: Any) -> bytes:
+    return msgpack.packb(obj, default=_msgpack_pack_extended_types)
+
+
+def _msgpack_unpack_extended_types(code: int, data: bytes) -> Any:
```
Comment (shcheklein) [inline src/datachain/query/queue.py:86]: can we please deduplicate this? (put into datachain utils, use in Studio)

File: src/datachain/query/queue.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+            tz_info = datetime.timezone(datetime.timedelta(seconds=timezone_offset))
+        return datetime.datetime.fromtimestamp(timestamp, tz=tz_info)
+
+    if code == MSGPACK_EXT_TYPE_ROWS_INPUT_BATCH:
+        return RowsInputBatch(msgpack_unpack(data))
+
+    return msgpack.ExtType(code, data)
+
+
+def msgpack_unpack(data: bytes) -> Any:
```
Comment (shcheklein) [inline src/datachain/query/queue.py:109]: why all of this is part of the `queue` file? seems like some general mspack stuff for sending inputs / output? or is exclusively used to move things through the queue?

File: src/datachain/query/dispatch.py (hunk 1 above)
Comment (dreadatour) [reply src/datachain/query/dispatch.py:56]: > are we moving things around in this case?

We do.

`dispatch.py` grow up quite a big after my initial updates and I was thinking about the way to reduce it's size and move out some parts to separate file (`queue.py`).

I can do it is separate PR though 👀

File: src/datachain/data_storage/warehouse.py (hunk 2 above)
Comment (dreadatour) [reply src/datachain/data_storage/warehouse.py:211]: It is more cosmetic and refactoring to simplify things.

File: src/datachain/data_storage/warehouse.py (hunk 3 above)
Comment (dreadatour) [reply src/datachain/data_storage/warehouse.py:238]: Yes, it is one of the major changes in this PR. There are three major things:

- return rows instead of dicts from SQL query (reduce memory consumption) **this one**
- marshal UDF rows with msgpack instead of dill (speedup pickling/types conversion)
- add batching to UDF results (speedup writing/reading to/from multiprocessing queues)

File: src/datachain/data_storage/warehouse.py (hunk 2 above)
Comment (shcheklein) [reply src/datachain/data_storage/warehouse.py:211]: sounds good, but please consider doing this things in a separate PR (can be a chained PR) ... or split by commits ... it makes it harder to review otherwise 🙏

File: src/datachain/utils.py (hunk 4 above)
Comment (dreadatour) [reply src/datachain/utils.py:265]: Nice catch, let me check 🙏

File: src/datachain/lib/udf.py (hunk 5 above)
Comment (dreadatour) [reply src/datachain/lib/udf.py:40]: Minor update, leftover, I can revert this change 👍

File: src/datachain/data_storage/warehouse.py (hunk 3 above)
Comment (shcheklein) [reply src/datachain/data_storage/warehouse.py:238]: can it be split into 3 PRs (chained or completely separate) + some cleanup? it would way easier to read, better history, less discussions on each :) 

(it's fine as it is now - I read almost all of it by now, but it took some time, and I feel we could have merged almost immediately some things, while others deserve some discussion)

File: src/datachain/query/dispatch.py (hunk 7 above)
Comment (dreadatour) [reply src/datachain/query/dispatch.py:333]: Hm, yes, you're right. I was doing this to stop typing warnings to pop up, but let me revert this 👍

File: src/datachain/query/dataset.py (hunk 8 above)
Comment (dreadatour) [reply src/datachain/query/dataset.py:522]: Before we do pass dicts to UDF as input data. Now we are passing tuples with data only, but we still need to know names of the fields, so we also passing `udf_fields` as list of names of the files (dicts keys in previous implementation). That's why we need it here. I can work a bit more on this because I see this may be improved a bit.

File: src/datachain/query/batch.py (hunk 9 above)
Comment (dreadatour) [reply src/datachain/query/batch.py:41]: Sounds good to me, let me recheck. Naming is hard 👀

File: src/datachain/query/batch.py (hunk 10 above)
Comment (dreadatour) [reply src/datachain/query/batch.py:75]: `execute` here is `dataset_select_paginated`, and it is related to this change here: https://github.com/iterative/datachain/pull/211/files/082757b233a2621a467c29f11a2181cd466ab6b8#diff-18ccbe3cf14ab41f123a6c5ec1848c31ddb1549e578ce4ce9db82622f09d994eL200-L206

Previously we run `dataset_select_paginated` with `limit` and `order_by` taken from `query` (it was a reason before, but now it is not). Now we are passing `query` only and take `limit` and `order_by` from query inside `dataset_select_paginated`.

File: src/datachain/query/queue.py (hunk 11 above)
Comment (dreadatour) [reply src/datachain/query/queue.py:86]: We do use it in Studio.
We are using `msgpack` in few places (like [here](https://github.com/iterative/datachain/blob/main/src/datachain/remote/studio.py#L145-L167)), and in each place we do have different set of extended types. I was thinking about putting it all together, but I am still not 100% convinced if this is a good idea to mix it all together 👀

File: src/datachain/query/queue.py (hunk 12 above)
Comment (dreadatour) [reply src/datachain/query/queue.py:109]: Yes, this exact msgpack pack/unpack with this set of extended types is used to move things through the queue. This is one of the reasons why I moved these code to the new file, btw.

File: src/datachain/data_storage/warehouse.py (hunk 2 above)
Comment (dreadatour) [reply src/datachain/data_storage/warehouse.py:211]: I know, I am trying to do my best but it is hard to resist sometimes :) Will isolate some changes to separate PR 👍

File: src/datachain/query/queue.py (hunk 12 above)
Comment (shcheklein) [reply src/datachain/query/queue.py:109]: is all these marshall / msgpack deal only with the queue though? ...they look more like generic helpers

File: src/datachain/data_storage/warehouse.py (hunk 3 above)
Comment (dreadatour) [reply src/datachain/data_storage/warehouse.py:238]: Let me think 👍

**Comment by dreadatour** [comment]
> Thank you, Vlad! Great improvements.

Thank you ❤️

> 1. **Is there a way to avoid the conversion?** We use a very limited subset of types. Why not just pass unflattened objects as tuple?
> 2. **Is there a way to avoid marshaling whatsoever?** We can write directly to DB.
> 
> Think of these as ideas for the next steps.

Yes, I have quite a lot of ideas to improve things more, these suggestions looks great! Let me put it all together in the optimizations epic 👍

File: src/datachain/query/queue.py (hunk 12 above)
Comment (dreadatour) [reply src/datachain/query/queue.py:109]: Yes, I see what you are talking about. It was the easiest way to do this, — to create one common set of `msgpack_pack`/`psgpack_unpack` functions. Let me think a bit more, I need to get used to this idea, it still feels a bit off to me personally.

File: src/datachain/utils.py (hunk 4 above)
Comment (rlamy) [reply src/datachain/utils.py:265]: When a generator returns, `StopIteration` is raised in the caller. Raising `StopIteration` inside the generator would be wrong, see https://peps.python.org/pep-0479/
